File: src/leetCode/array/Problem033.java
package leetCode.array;

/**
 * Suppose an array sorted in ascending order is rotated at some pivot unknown to you beforehand.
 * (i.e., [0,1,2,4,5,6,7] might become [4,5,6,7,0,1,2]).
 * You are given a target value to search. If found in the array return its index, otherwise return -1.
 * You may assume no duplicate exists in the array.
 * Your algorithm's runtime complexity must be in the order of O(log n).
 * <p>
 * Example 1:
 * Input: nums = [4,5,6,7,0,1,2], target = 0
 * Output: 4
 * <p>
 * Example 2:
 * Input: nums = [4,5,6,7,0,1,2], target = 3
 * Output: -1
 * <p>
 * create by stephen on 2018/5/4
 */
public class Problem033 {

    /**
     * 二分查找 只要找到子结构即可
     */
    public int search(int[] nums, int target) {
        if (nums.length == 0) return -1;
        if (nums[0] == target) return 0;
        int left = 0, right = nums.length - 1, mid;
        while (left < right) {
            if (nums[left] == target) return left;
            if (nums[right] == target) return right;
            mid = (left + right) / 2;
            if (nums[mid] == target) return mid;

            if (nums[left] < target) { //target在数字大的半边
                if (nums[mid] > nums[left]) {       //中间数字在大的半边
                    if (nums[mid] > target) {
                        right = mid;
                    } else {
                        left = mid;
                    }
                } else {                    //中间数字在小的半边
                    right = mid;
                }
            } else {            //target在数字小的半边
                if (nums[mid] > nums[left]) {       //中间数字在大的半边
                    left = mid;
                } else {                    //中间数字在小的半边
                    if (nums[mid] < target) {
                        left = mid;
                    } else {
                        right = mid;
                    }
                }
            }
        }

        return -1;
    }
}

File: leetCode/binSearch/Problem033.cpp

#include <climits>
#include <vector>

/**
 * You are given an integer array nums sorted in ascending order, and an integer
 * target.
 *
 * Suppose that nums is rotated at some pivot unknown to you beforehand (i.e.,
 * [0,1,2,4,5,6,7] might become [4,5,6,7,0,1,2]).
 *
 * If target is found in the array return its index, otherwise, return -1.
 */
int search(std::vector<int>& nums, int target) {
  int start = 0, end = nums.size();
  // need to consider four cases according to the relation between
  // nums[start], nums[end], nums[mid], and target
  while (start < end) {
    int m = (start + end) >> 1;
    if (nums[start] > nums[m]) {
      if (target < nums[m])
        end = m;
      else {
        if (target >= nums[start])
          end = m;
        else
          start = m;
      }
    } else if (nums[start] < nums[m]) {
      if (target >= nums[m])
        start = m;
      else {
        if (target >= nums[start])
          end = m;
        else
          start = m + 1;
      }
    } else {
      break;
    }
  }
  return (start < nums.size() && nums[start] == target) ? start : -1;
}